File: EJEMPLOS PATRONES/PHP/TemplateMethod.php
<?php

abstract class Vehicle {

    public abstract function start();

    public abstract function run();

    public abstract function stop();

    public function testYourVehicle() {
        $this->start();
        $this->run();
        $this->stop();
    }

}

class Car extends Vehicle {
    private function start() {
        
    }
    private function run() {
        
    }
    private function stop() {
        
    }

}

class Truck extends Vehicle {
    private function start() {
        
    }
    private function run() {
        
    }
    private function stop() {
        
    }
}

class Tester {
    public function testVehicle() {
        $car = new Car();
        $truck = new Truck();
        
        $car->testYourVehicle();
        $truck->testYourVehicle();
    }

}

File: EJEMPLOS PATRONES/Python/TemplateMethod.py
import abc


class Vehicle(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def _start(self):
        pass

    @abc.abstractmethod
    def _stop(self):
        pass

    @abc.abstractmethod
    def _run(self):
        pass

    def testYourVehicle(self):
        self._start()
        self._run()
        self._stop()


class Car(Vehicle):
    def _start(self):
        pass

    def _stop(self):
        pass

    def _run(self):
        pass


class Truck(Vehicle):
    def _start(self):
        pass

    def _stop(self):
        pass

    def _run(self):
        pass


def main():
    vehicle = Car()
    vehicle.testYourVehicle()
